File: AndroidUtils/src/main/java/com/github/teranes10/androidutils/utils/FtpUtil.java
package com.github.teranes10.androidutils.utils;

import android.os.Environment;
import android.util.Log;

import org.apache.commons.net.ftp.FTP;
import org.apache.commons.net.ftp.FTPClient;
import org.apache.commons.net.ftp.FTPFile;
import org.apache.commons.net.io.CopyStreamAdapter;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.concurrent.CompletableFuture;

import au.com.softclient.mydevices.models.response.Result;

public class FtpUtil {
    private static final String TAG = "FtpUtil";

    public static CompletableFuture<Result<?>> downloadDir(FtpDirConfig config, File toDir, FtpDirTransferListener listener) {
        return CompletableFuture.supplyAsync(() -> {
            FTPClient ftp = new FTPClient();
            OutputStream outputStream = null;
            try {
                if (!isExternalStorageWritable()) {
                    Log.e(TAG, "downloadDir: Write access not granted.");
                    return Result.fail("Write access not granted.");
                }

                ftp.connect(config.server, config.port);
                Log.d(TAG, "Connected.");

                ftp.login(config.user, config.password);
                Log.d(TAG, "Logged in");

                //    ftp.setFileType(FTP.BINARY_FILE_TYPE);
                ftp.enterLocalPassiveMode();

                ftp.changeWorkingDirectory(config.contentRoot);
                showServerReply(ftp);

                if (!toDir.exists()) {
                    boolean created = toDir.mkdirs();
                    if (created) {
                        Log.i(TAG, "downloadDir: Directory Created.");
                    }
                }


                FTPFile[] files = ftp.listFiles(config.dir);
                if (files == null || files.length == 0) {
                    Log.e(TAG, "downloadDir: No Files.");
                    return Result.fail("No files.");
                }

                ftp.changeWorkingDirectory(config.contentRoot + "/" + config.dir);

                final int filesTotal = files.length;
                int[] downloadedFilesTotal = new int[]{0};
                long[] fileSize = new long[]{0};

                if (listener != null) {
                    ftp.setCopyStreamListener(new CopyStreamAdapter() {

                        @Override
                        public void bytesTransferred(long totalBytesTransferred, int bytesTransferred, long streamSize) {
                            if (fileSize[0] > 0) {
                                int percent = (int) (totalBytesTransferred * 100 / fileSize[0]);
                                listener.onChanged(percent, downloadedFilesTotal[0], filesTotal);
                            }
                        }
                    });
                }

                for (FTPFile file : files) {
                    fileSize[0] = file.getSize();
                    File localFile = new File(toDir.getAbsolutePath() + "/" + file.getName());
                    outputStream = new BufferedOutputStream(new FileOutputStream(localFile));
                    boolean success = ftp.retrieveFile(file.getName(), outputStream);
                    if (success) {
                        ++downloadedFilesTotal[0];
                        if (listener != null) {
                            listener.onChanged(100, downloadedFilesTotal[0], filesTotal);
                        }
                    }
                }

                if (filesTotal == downloadedFilesTotal[0]) {
                    Log.i(TAG, "downloadFile: Downloaded successfully.");
                    return Result.ok("Downloaded successfully.");
                }
            } catch (IOException e) {
                Log.e(TAG, "downloadFile: ", e);
                return Result.fail("Something went wrong. " + e.getLocalizedMessage());
            } finally {
                try {
                    ftp.logout();
                    ftp.disconnect();
                    if (outputStream != null) {
                        outputStream.close();
                    }
                } catch (IOException e) {
                    Log.e(TAG, "downloadFile: ", e);
                }
            }

            return Result.fail("Something went wrong.");
        }).exceptionally(e -> Result.fail("Something went wrong."));
    }

    public static CompletableFuture<Result<Object>> downloadFile(FtpFileConfig config, File localFile) {
        return downloadFile(config, localFile);
    }

    public static CompletableFuture<Result<?>> downloadFile(FtpFileConfig config, File localFile, FtpFileTransferListener listener) {
        return CompletableFuture.supplyAsync(() -> {
            FTPClient ftp = new FTPClient();
            OutputStream outputStream = null;
            try {
                if (!isExternalStorageWritable()) {
                    Log.e(TAG, "downloadFile: Write access not granted.");
                    return Result.fail("Write access not granted.");
                }

                ftp.connect(config.server, config.port);
                Log.d(TAG, "Connected.");

                ftp.login(config.user, config.password);
                Log.d(TAG, "Logged in");

                ftp.setFileType(FTP.BINARY_FILE_TYPE);
                ftp.enterLocalPassiveMode();

                ftp.changeWorkingDirectory(config.contentRoot);
                showServerReply(ftp);

                String dirPath = localFile.getParent();
                if (dirPath != null) {
                    File dir = new File(dirPath);
                    if (!dir.exists()) {
                        boolean created = dir.mkdirs();
                        if (created) {
                            Log.i(TAG, "downloadFile: Directory Created.");
                        }
                    }
                }

                FTPFile[] files = ftp.listFiles(config.fileName);
                if (files == null || files.length == 0) {
                    Log.e(TAG, "downloadFile: File not found.");
                    return Result.fail("File not found.");
                }

                long fileSize = files[0].getSize();

                CopyStreamAdapter streamListener = new CopyStreamAdapter() {

                    @Override
                    public void bytesTransferred(long totalBytesTransferred, int bytesTransferred, long streamSize) {
                        int percent = (int) (totalBytesTransferred * 100 / fileSize);
                        if (listener != null) {
                            listener.onChanged(percent);
                        }
                    }
                };

                ftp.setCopyStreamListener(streamListener);
                outputStream = new BufferedOutputStream(new FileOutputStream(localFile));
                boolean success = ftp.retrieveFile(config.fileName, outputStream);
                if (success) {
                    Log.i(TAG, "downloadFile: Downloaded successfully.");
                    return Result.ok("Downloaded successfully.");
                }
            } catch (IOException e) {
                Log.e(TAG, "downloadFile: ", e);
                return Result.fail("Something went wrong. " + e.getLocalizedMessage());
            } finally {
                try {
                    ftp.logout();
                    ftp.disconnect();
                    if (outputStream != null) {
                        outputStream.close();
                    }
                } catch (IOException e) {
                    Log.e(TAG, "downloadFile: ", e);
                }
            }

            return Result.fail("Something went wrong.");
        }).exceptionally(e -> Result.fail("Something went wrong."));
    }

    private static void showServerReply(FTPClient ftpClient) {
        String[] replies = ftpClient.getReplyStrings();
        if (replies != null) {
            for (String aReply : replies) {
                Log.i(TAG, "showServerReply: " + aReply);
            }
        }
    }

    private static boolean isExternalStorageWritable() {
        String state = Environment.getExternalStorageState();
        return Environment.MEDIA_MOUNTED.equals(state);
    }

    public static class FtpDirConfig extends FtpConfig {
        public String dir;

        public FtpDirConfig(String server, int port, String user, String password, String contentRoot, String dir) {
            super(server, port, user, password, contentRoot);
            this.dir = dir;
        }
    }

    public static class FtpFileConfig extends FtpConfig {
        public String fileName;

        public FtpFileConfig(String server, int port, String user, String password, String contentRoot, String fileName) {
            super(server, port, user, password, contentRoot);
            this.fileName = fileName;
        }
    }

    private static class FtpConfig {
        public String server;
        public int port;
        public String user;
        public String password;
        public String contentRoot;

        public FtpConfig(String server, int port, String user, String password, String contentRoot) {
            this.server = server;
            this.port = port;
            this.user = user;
            this.password = password;
            this.contentRoot = contentRoot;
        }
    }

    public interface FtpFileTransferListener {
        void onChanged(int percentage);
    }

    public interface FtpDirTransferListener {
        void onChanged(int percentage, int filesDownloaded, int totalFiles);
    }
}

File: app/src/main/java/com/github/teranes10/androidutils/utils/FtpUtil.kt
package com.github.teranes10.androidutils.utils

import android.os.Environment
import android.util.Log
import com.github.teranes10.androidutils.models.Outcome
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.apache.commons.net.ftp.*
import org.apache.commons.net.io.CopyStreamAdapter
import java.io.*

object FtpUtil {
    private const val TAG = "FtpUtil"

    suspend fun downloadDir(config: FtpDirConfig, toDir: File, listener: FtpDirTransferListener? = null): Outcome<*> =
        withContext(Dispatchers.IO) {
            val ftp = FTPClient()
            try {
                if (!isExternalStorageWritable()) {
                    Log.e(TAG, "downloadDir: Write access not granted.")
                    return@withContext Outcome.fail(false, "Write access not granted.")
                }

                ftp.connect(config.server, config.port)
                Log.d(TAG, "Connected.")

                ftp.login(config.user, config.password)
                Log.d(TAG, "Logged in.")

                ftp.enterLocalPassiveMode()
                ftp.changeWorkingDirectory(config.contentRoot)
                showServerReply(ftp)

                if (!toDir.exists() && toDir.mkdirs()) {
                    Log.i(TAG, "downloadDir: Directory Created.")
                }

                val files = ftp.listFiles(config.dir)
                if (files.isNullOrEmpty()) {
                    Log.e(TAG, "downloadDir: No Files.")
                    return@withContext Outcome.fail(false, "No files.")
                }

                ftp.changeWorkingDirectory("${config.contentRoot}/${config.dir}")

                val totalFiles = files.size
                var downloadedFiles = 0
                var fileSize = 0L

                listener?.let {
                    ftp.copyStreamListener = object : CopyStreamAdapter() {
                        override fun bytesTransferred(totalBytesTransferred: Long, bytesTransferred: Int, streamSize: Long) {
                            if (fileSize > 0) {
                                val percent = (totalBytesTransferred * 100 / fileSize).toInt()
                                listener.onChanged(percent, downloadedFiles, totalFiles)
                            }
                        }
                    }
                }

                for (file in files) {
                    fileSize = file.size
                    val localFile = File(toDir, file.name)
                    BufferedOutputStream(FileOutputStream(localFile)).use { outputStream ->
                        if (ftp.retrieveFile(file.name, outputStream)) {
                            downloadedFiles++
                            listener?.onChanged(100, downloadedFiles, totalFiles)
                        }
                    }
                }

                if (downloadedFiles == totalFiles) {
                    Log.i(TAG, "downloadDir: Downloaded successfully.")
                    return@withContext Outcome.ok(true, "Downloaded successfully.")
                }

                Outcome.fail(false, "Some files failed to download.")
            } catch (e: IOException) {
                Log.e(TAG, "downloadDir: ", e)
                Outcome.fail(false, e.localizedMessage ?: e.message ?: "Something went wrong.")
            } finally {
                ftp.logout()
                ftp.disconnect()
            }
        }

    suspend fun downloadFile(config: FtpFileConfig, localFile: File, listener: FtpFileTransferListener? = null): Outcome<*> =
        withContext(Dispatchers.IO) {
            val ftp = FTPClient()
            try {
                if (!isExternalStorageWritable()) {
                    Log.e(TAG, "downloadFile: Write access not granted.")
                    return@withContext Outcome.fail(false, "Write access not granted.")
                }

                ftp.connect(config.server, config.port)
                Log.d(TAG, "Connected.")

                ftp.login(config.user, config.password)
                Log.d(TAG, "Logged in.")

                ftp.setFileType(FTP.BINARY_FILE_TYPE)
                ftp.enterLocalPassiveMode()
                ftp.changeWorkingDirectory(config.contentRoot)
                showServerReply(ftp)

                localFile.parentFile?.takeIf { !it.exists() }?.mkdirs()

                val files = ftp.listFiles(config.fileName)
                if (files.isNullOrEmpty()) {
                    Log.e(TAG, "downloadFile: File not found.")
                    return@withContext Outcome.fail(false, "File not found.")
                }

                val fileSize = files[0].size

                ftp.copyStreamListener = object : CopyStreamAdapter() {
                    override fun bytesTransferred(totalBytesTransferred: Long, bytesTransferred: Int, streamSize: Long) {
                        val percent = (totalBytesTransferred * 100 / fileSize).toInt()
                        listener?.onChanged(percent)
                    }
                }

                BufferedOutputStream(FileOutputStream(localFile)).use { outputStream ->
                    if (ftp.retrieveFile(config.fileName, outputStream)) {
                        Log.i(TAG, "downloadFile: Downloaded successfully.")
                        return@withContext Outcome.ok(true, "Downloaded successfully.")
                    }
                }

                Outcome.fail(false, "Download failed.")
            } catch (e: IOException) {
                Log.e(TAG, "downloadFile: ", e)
                Outcome.fail(false, e.localizedMessage ?: e.message ?: "Something went wrong.")
            } finally {
                ftp.logout()
                ftp.disconnect()
            }
        }

    private fun showServerReply(ftpClient: FTPClient) {
        ftpClient.replyStrings?.forEach { reply ->
            Log.i(TAG, "showServerReply: $reply")
        }
    }

    private fun isExternalStorageWritable(): Boolean {
        return Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED
    }

    data class FtpDirConfig(
        val server: String,
        val port: Int,
        val user: String,
        val password: String,
        val contentRoot: String,
        val dir: String
    )

    data class FtpFileConfig(
        val server: String,
        val port: Int,
        val user: String,
        val password: String,
        val contentRoot: String,
        val fileName: String
    )

    fun interface FtpFileTransferListener {
        fun onChanged(percentage: Int)
    }

    fun interface FtpDirTransferListener {
        fun onChanged(percentage: Int, filesDownloaded: Int, totalFiles: Int)
    }
}
